packages/hash/shared/src/save.ts

```diff
@@ -312,11 +316,64 @@ const updateBlocks = defineOperation(
                 },
               });
             }
+          } else {
+            const draftBlock = getDraftEntityFromEntityId(
+              entityStore.draft,
+              savedEntity.entityId,
+            );
+            if (isBlockEntity(draftBlock)) {
+              const draftChildEntityId =
+                draftBlock?.properties.entity?.entityId;
+              const draftChildEntityWithDraftId = getDraftEntityFromEntityId(
+                entityStore.draft,
+                draftChildEntityId,
+              );
+
+              if (draftChildEntityWithDraftId && draftChildEntityId) {
+                const { draftId: _, ...draftChildEntity } =
+                  draftChildEntityWithDraftId;
+
+                // Check if a block's data has been updated
+                if (
+                  draftChildEntity &&
+                  !isEqual(
+                    draftChildEntity.properties,
+                    savedChildEntity.properties,
+                  )
+                ) {
+                  // If the block's data was updated and the entityId changed
+                  // then the block data was swapped. Push swapBlockData action
+                  if (draftChildEntity.entityId !== savedChildEntity.entityId) {
+                    updates.push({
+                      swapBlockData: {
+                        entityId: savedEntity.entityId,
+                        accountId: savedEntity.accountId,
+                        newEntityAccountId: draftChildEntity.accountId!,
+                        newEntityEntityId: draftChildEntity.entityId!,
+                      },
+                    });
+                  } else {
+                    // If the block's data was updated and the entityId
+                    // didn't change, push an updateEntity action
+                    // @todo push an updateEntity action
```

Comment:
Yeah this will be necessary – but to be honest when we add this its probably worth rewriting this whole function to remove the text block specific changes. In reality the whole thing needs rewriting to work from the entity store and not the prosemirror tree.

Reply:
Added a comment for this here => https://github.com/hashintel/hash/pull/506/commits/276fd19b917f77f590cb8932f07891b46225244a